core/src/main/java/io/grpc/Metadata.java

```diff
@@ -235,7 +305,25 @@ public boolean containsKey(Key<?> key) {
   public <T> void put(Key<T> key, T value) {
     Preconditions.checkNotNull(key, "key");
     Preconditions.checkNotNull(value, "value");
-    storeAdd(key.name, new MetadataEntry(key, value));
+    if (len() == 0 || len() == cap()) {
+      int desiredCap = len() + (len() >>> 1) + 4;
```

Comment:
why the complexity and not just `desiredCap  = max(8, len() * 2)` or so. Metadata is not expected to grow large (i.e. thousands or millions of entries).